public/app/core/utils/richHistory.ts

```diff
@@ -46,164 +42,143 @@ export function addToRichHistory(
-      }
-      current--;
-    }
-
-    let updatedHistory: RichHistoryQuery[] = [
-      {
-        queries: newQueriesToSave,
-        ts,
-        datasourceId,
-        datasourceName: datasourceName ?? '',
-        starred,
-        comment: comment ?? '',
-        sessionName,
-      },
-      ...queriesToKeep,
-    ];
+    let warning: RichHistoryStorageWarningDetails | undefined;
 
     try {
-      showLimitExceededWarning &&
-        limitExceeded &&
-        dispatch(
-          notifyApp(
-            createWarningNotification(
-              `Query history reached the limit of ${MAX_HISTORY_ITEMS}. Old, not-starred items will be removed.`
-            )
-          )
-        );
-      store.setObject(RICH_HISTORY_KEY, updatedHistory);
-      return { richHistory: updatedHistory, limitExceeded, localStorageFull: false };
+      warning = await getRichHistoryStorage().addToRichHistory(newRichHistory);
     } catch (error) {
-      showQuotaExceededError &&
-        dispatch(notifyApp(createErrorNotification('Saving rich history failed', error.message)));
-      return { richHistory: updatedHistory, limitExceeded, localStorageFull: error.name === 'QuotaExceededError' };
+      if (error.name === RichHistoryServiceError.StorageFull) {
+        richHistoryStorageFull = true;
+        showQuotaExceededError && dispatch(notifyApp(createErrorNotification(error.message)));
+      } else if (error.name !== RichHistoryServiceError.DuplicatedEntry) {
+        dispatch(notifyApp(createErrorNotification('Rich History update failed', error.message)));
+      }
+      // Saving failed. Do not add new entry.
+      return { richHistory, richHistoryStorageFull, limitExceeded };
+    }
+
+    // Limit exceeded but new entry was added. Notify that old entries have been removed.
+    if (warning && warning.type === RichHistoryStorageWarning.LimitExceeded) {
+      limitExceeded = true;
+      showLimitExceededWarning && dispatch(notifyApp(createWarningNotification(warning.message)));
     }
+
+    // Saving successful - add new entry.
+    return { richHistory: [newRichHistory, ...richHistory], richHistoryStorageFull, limitExceeded };
   }
 
+  // Nothing to save
   return { richHistory };
 }
 
-export function getRichHistory(): RichHistoryQuery[] {
-  const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
-  const transformedRichHistory = migrateRichHistory(richHistory);
-  return transformedRichHistory;
+export async function getRichHistory(): Promise<RichHistoryQuery[]> {
+  return await getRichHistoryStorage().getRichHistory();
 }
 
-export function deleteAllFromRichHistory() {
-  return store.delete(RICH_HISTORY_KEY);
+export async function deleteAllFromRichHistory(): Promise<void> {
+  return getRichHistoryStorage().deleteAll();
 }
 
-export function updateStarredInRichHistory(richHistory: RichHistoryQuery[], ts: number) {
+export async function updateStarredInRichHistory(richHistory: RichHistoryQuery[], ts: number) {
```

Comment:
i'm not 100% sure the functions below should return a cached version of the history instead of requesting a fresh one from the rich history storage. i can see why we want to do it, but while it may make sense when using localstorage, i think this could be a bit more error prone when using a remote storage. wdyt?

Reply:
This will need to change, I just didn't want to change the current behavior too much.

We need to take into account error handling. If starring/commenting/deleting fails we should not apply changes in frontend (this may happen when one user deletes an entry and someone else tries to edit). I'll add it to do list.